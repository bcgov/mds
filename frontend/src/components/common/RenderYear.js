/* eslint-disable */
import React, { Component } from "react";
import PropTypes from "prop-types";
import { Form, DatePicker } from "antd";
import moment from "moment";

/**
 * @constant RenderDate  - Ant Design `DatePicker` component for redux-form.
 */

const propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  input: PropTypes.objectOf(PropTypes.any).isRequired,
  label: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  onChange: PropTypes.func,
  disabled: PropTypes.bool,
  meta: PropTypes.objectOf(PropTypes.any).isRequired,
};

const defaultProps = {
  placeholder: "yyyy-mm-dd",
  onChange: () => {},
  disabled: false,
};

export class RenderDate extends Component {
  state = {
    isopen: false,
    time: this.props.input.value ? moment(`${this.props.input.value}-01-01`) : null,
  };

  componentWillReceiveProps = (nextProps) => {
    this.setState({
      time: nextProps.input.value ? moment(`${nextProps.input.value}-01-01`) : null,
    });
  };

  handlePanelChange = (value) => {
    this.setState({
      time: value,
      isopen: false,
    });
    this.props.input.onChange(moment(value).format("YYYY"));
  };

  handleOpenChange = (status) => {
    if (status) {
      this.setState({ isopen: true });
    } else {
      this.setState({ isopen: false });
    }
  };

  clearValue = () => {
    this.setState({
      time: null,
    });
  };

  render = () => (
    <Form.Item
      label={this.props.label}
      validateStatus={
        this.props.meta.touched
          ? (this.props.meta.error && "error") || (this.props.meta.warning && "warning")
          : ""
      }
      help={
        this.props.meta.touched &&
        ((this.props.meta.error && <span>{this.props.meta.error}</span>) ||
          (this.props.meta.warning && <span>{this.props.meta.warning}</span>))
      }
    >
      <DatePicker
        id={this.props.id}
        value={this.state.time}
        open={this.state.isopen}
        placeholder={this.props.placeholder}
        mode="year"
        format="YYYY"
        onOpenChange={this.handleOpenChange}
        onPanelChange={this.handlePanelChange}
        onChange={this.clearValue}
        disabled={this.props.disabled}
      />
    </Form.Item>
  );
}

RenderDate.propTypes = propTypes;
RenderDate.defaultProps = defaultProps;

export default RenderDate;
