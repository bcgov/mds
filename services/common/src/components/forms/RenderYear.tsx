import React, { Component } from "react";
import { Form } from "@ant-design/compatible";
import "@ant-design/compatible/assets/index.css";
import { DatePicker } from "antd";
import moment, { Moment } from "moment";

/**
 * @constant RenderYear  - Ant Design `DatePicker` component for redux-form.
 */

interface RenderYearProps {
  id: any;
  input: Record<string, any>;
  label: string;
  placeholder?: string;
  required?: boolean;
  onChange?: (date: Moment | null, dateString: string) => void;
  disabledDate?: (current: Moment | null) => boolean;
  disabled?: boolean;
  meta: Record<string, any>;
}

const defaultProps = {
  placeholder: "",
  onChange: () => {},
  disabledDate: () => false, // Default value for disabledDate
  disabled: false,
  required: false,
};

type RenderYearState = {
  isopen: boolean;
  time: Moment | null;
};

export class RenderYear extends Component<RenderYearProps, RenderYearState> {
  static defaultProps = defaultProps; // Initialize default props here

  constructor(props: RenderYearProps) {
    super(props);
    this.state = {
      isopen: false,
      time: props.input.value ? moment(`${props.input.value}-01-01`) : null,
    };
  }

  handlePanelChange = (value: Moment) => {
    this.setState({
      time: value,
      isopen: false,
    });
    this.props.onChange?.(value, moment(value).format("YYYY"));
  };

  handleOpenChange = (status: boolean) => {
    this.setState({ isopen: status });
  };

  clearValue = () => {
    this.setState({
      time: null,
    });
  };

  render = () => (
    <Form.Item
      label={this.props.label}
      validateStatus={
        this.props.meta.touched
          ? (this.props.meta.error && "error") || (this.props.meta.warning && "warning")
          : ""
      }
      help={
        this.props.meta.touched &&
        ((this.props.meta.error && <span>{this.props.meta.error}</span>) ||
          (this.props.meta.warning && <span>{this.props.meta.warning}</span>))
      }
      required={this.props.required}
    >
      <DatePicker
        id={this.props.id}
        value={this.state.time}
        open={this.state.isopen}
        placeholder={this.props.placeholder}
        mode="year"
        format="YYYY"
        onOpenChange={this.handleOpenChange}
        onPanelChange={this.handlePanelChange}
        onChange={this.clearValue}
        disabledDate={this.props.disabledDate}
        disabled={this.props.disabled}
      />
    </Form.Item>
  );
}

export default RenderYear;
