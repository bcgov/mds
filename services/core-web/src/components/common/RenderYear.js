/* eslint-disable */
import React, { Component } from "react";
import PropTypes from "prop-types";
import { Form, DatePicker } from "antd";
import moment from "moment";

/**
 * @constant RenderDate  - Ant Design `DatePicker` component for redux-form.
 */

const propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  input: PropTypes.objectOf(PropTypes.any).isRequired,
  meta: PropTypes.objectOf(PropTypes.any).isRequired,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
};

const defaultProps = {
  label: "",
  placeholder: "",
  disabled: false,
};

const getInputValue = (value) => (value ? moment(`${value}-01-01`) : null);

export class RenderDate extends Component {
  state = {
    value: getInputValue(this.props.input.value),
    isOpen: false,
  };

  componentWillReceiveProps = (nextProps) => {
    if (nextProps.input.value !== this.props.input.value) {
      this.setState({
        value: getInputValue(nextProps.input.value),
      });
    }
  };

  handlePanelChange = (value) => {
    this.setState({
      value: value,
      isOpen: false,
    });
    this.props.input.onChange(value ? moment(value).format("YYYY") : null);
  };

  handleOpenChange = (isOpen) => {
    this.setState({ isOpen });
  };

  handleChange = () => {
    this.setState({
      value: null,
    });
    this.props.input.onChange(null);
  };

  render = () => (
    <Form.Item
      label={this.props.label}
      validateStatus={
        this.props.meta.touched
          ? (this.props.meta.error && "error") || (this.props.meta.warning && "warning")
          : ""
      }
      help={
        this.props.meta.touched &&
        ((this.props.meta.error && <span>{this.props.meta.error}</span>) ||
          (this.props.meta.warning && <span>{this.props.meta.warning}</span>))
      }
    >
      <DatePicker
        id={this.props.id}
        value={this.state.value}
        open={this.state.isOpen}
        placeholder={this.props.placeholder}
        disabled={this.props.disabled}
        mode="year"
        format="YYYY"
        onOpenChange={this.handleOpenChange}
        onPanelChange={this.handlePanelChange}
        onChange={this.handleChange}
      />
    </Form.Item>
  );
}

RenderDate.propTypes = propTypes;
RenderDate.defaultProps = defaultProps;

export default RenderDate;
