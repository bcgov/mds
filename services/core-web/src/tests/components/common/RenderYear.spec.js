import React from "react";
import { shallow } from "enzyme";
import RenderYear from "@/components/common/RenderYear";

let props = {};

const setupProps = () => {
  props = {
    id: 1,
    input: { value: "" },
    label: "Time",
    meta: {
      touched: false,
      error: false,
      warning: false,
    },
  };
};

beforeEach(() => {
  setupProps();
});

describe("RenderYear", () => {
  it("renders properly", () => {
    const wrapper = shallow(<RenderYear {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
