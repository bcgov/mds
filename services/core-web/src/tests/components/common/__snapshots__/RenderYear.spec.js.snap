// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderYear renders properly 1`] = `
<FormItem
  hasFeedback={false}
  help={false}
  label="Time"
  validateStatus=""
>
  <Picker
    disabled={false}
    format="YYYY"
    id={1}
    mode="year"
    onChange={[Function]}
    onOpenChange={[Function]}
    onPanelChange={[Function]}
    open={false}
    placeholder=""
    value={null}
  />
</FormItem>
`;
